### ISSUE TITLE
Decide whether the reverse-engineered generator is worth using

### ISSUE THREAD

**Issue by ronaldoi9** [ronaldoi9/predicting_vehicle_purchases#5]
## Question

Is the reverse-engineered generator worth using — as `init_score`, as features, or at all?

The competition research (#3) established that Chris Deotte reverse-engineered this dataset's generator: **RNG seed 101 recovered**, the original 10,000 rows reproduced exactly, and the label rule published as

```
buy_score = 1.2*(income/1e5) + 0.6*concern + 2*subsidy - 1*(anxiety == medium) - 3*(anxiety == high) + noise
positive if buy_score > 5.5
```

plus city-conditional charger ranges and home-charging probabilities, and a hard income floor at $30,000 (the profile in #2 confirms 9.21% of rows sit exactly on that floor).

External data **is** allowed (Rules 2.6a), so nothing here is blocked on permission.

Decide:

- **Does the recipe reproduce our data?** Score every train row with the published formula and measure the resulting AUC directly. This is a read-only check and it either validates the recipe or shows it is incomplete — both are useful. If the formula alone approaches the plateau, the residual is noise and the ceiling is arithmetic rather than a modelling problem.
- **Recipe as `init_score` / `base_margin`.** The field measured **+0.00005 to +0.00007**, replicated across XGBoost and LightGBM, and near-zero when the same quantity is passed as an ordinary feature. That is above the 0.00002 paired-noise floor but barely. Decide whether it earns a place in turn 1 or waits — it is cheap, but it adds a moving part to the baseline.
- **The original 10,000 rows as extra training data.** 10k against 668k is a 1.5% row bump with a matching distribution, and the field measured label-joining at +0.00003. State plainly whether to rule this out.
- **Is reproducing the generator ourselves worth the time**, or do we take the published formula as given and verify it empirically against our train set?

## Why it matters

This decides whether the generator is a tool or a distraction. If the recipe explains the data as well as the field claims, it sets a hard ceiling that makes the gap-strategy decision straightforward. Note the shape of the evidence: the two biggest single-feature AUCs in #2 (`Environmental_Concern_Level` 0.844, `Subsidy_Available` 0.718) are exactly the two largest coefficients in the published formula, which is strong independent corroboration.

**Comment by ronaldoi9** [comment]
## Findings: the recipe is real, but it is NOT the ceiling

Full document: [`docs/research/generator-recipe.md`](https://github.com/ronaldoi9/predicting_vehicle_purchases/blob/research/generator-recipe/docs/research/generator-recipe.md) on branch `research/generator-recipe` (not merged). Read-only analysis plus throwaway LightGBM fits used only as measurement instruments — no pipeline code, no model kept.

### The central measurement

Every one of the 668,665 train rows scored with the published formula, ranked by the **noiseless** deterministic part (additive independent noise makes `P(y=1|x)` monotone in `buy_score`, so the noiseless score *is* the Bayes ranking under the published model):

| quantity | AUC |
|---|---|
| **published formula, noiseless, as a ranking** | **0.937690** |
| free-coefficient logistic refit on the same 4 variables | 0.937772 |
| published plateau | ~0.9464 CV / 0.9466 LB |
| **gap** | **−0.0089** |

**The ticket's headline hypothesis is refuted.** The formula does *not* land near the plateau. 0.0089 — about 45× the LB noise floor of 0.0002 — separates the published recipe from what the field routinely reaches. The residual is **not** noise; this project's ceiling is a modelling problem, not arithmetic.

**But the recipe itself is confirmed.** A free logistic refit on exactly the recipe's variables, renormalised so concern = 0.6, recovers: income **1.157** (published 1.2), subsidy **2.072** (2), anxiety Medium **−0.841** (−1), High **−1.728** (−3, but High is 0.33% of rows so weakly identified), threshold **5.647** (5.5). The refit buys **+0.00008 AUC**. The published constants are right to the fifth decimal of AUC.

### Where the missing 0.0089 lives

5-fold `StratifiedKFold`, plain LightGBM baseline (no digit decomposition, no target encoding — absolute level sits below the plateau; the deltas are the point):

| model | OOF AUC | delta |
|---|---|---|
| published formula as a score (no model) | 0.937690 | — |
| LightGBM on the **4 recipe variables only** | **0.940632** | **+0.00294** |
| LightGBM on all 13 features | 0.941665 | +0.00103 |
| published FE on top (digit +0.00143, triple TE +0.00129, freq +0.00112) | — | ~+0.0038 |

**Two thirds of the gap is inside the recipe's own four variables, used non-linearly.** A GBDT given nothing but income, concern, subsidy and anxiety beats the linear recipe on those same columns by +0.0029, while the free *linear* refit bought +0.00008. So the label in our 668k rows is not a monotone function of any linear index of the recipe inputs — which is the field's exact-income-artifact finding seen from the other side, and explains *why* digit decomposition and income target encoding pay. Only +0.0010 comes from the nine features the formula never mentions.

The arithmetic closes: 0.9377 + 0.0029 + 0.0010 = 0.9417 (measured) + ~0.0038 published FE ≈ 0.9455 + tuning ≈ the plateau.

**Interpretation:** the published recipe is the label rule of the **original 10,000 rows**. Our 668k were regenerated by a learned generative model that only approximately preserves it. Treat the formula as a description of the *source*, not of the competition data's DGP.

### Sanity checks (all three requested, all informative)

**1. Positive rate — nearly, not exactly.** `buy_score > 5.5` with zero noise flags **17.794%** vs the observed **17.4645%**. Adding symmetric zero-mean noise only raises the rate (0.1781 at sigma 0.01 → 0.2012 at sigma 1.0 → 0.2671 at sigma 2.0), so **no noise scale reproduces 17.4645% with a 5.5 threshold** — infimum 0.1778. The threshold would have to be ~5.65, exactly what the free refit recovered. **Inferred noise: Gaussian sigma = 0.802** (MLE, mean log-loss 0.23622); logistic scale 0.444 fits marginally better (0.23610) and is indistinguishable. A plain `N(0,1)` is rejected.

**2. Coefficients vs single-feature AUCs — they match.** SD of each term's contribution to `buy_score`: subsidy 0.967 (AUC 0.718), concern 0.858 (**0.844**), income 0.344 (0.670), anxiety 0.335 (0.545). Ordering matches once granularity is accounted for — subsidy has the widest spread but is binary and AUC-capped by prevalence; income and anxiety have equal spread but anxiety is 90.3% Low / 9.3% Medium / **0.33% High**, so the −3 term touches almost nobody. Independent corroboration from a direction the recipe was not fitted on.

**3. Systematic misprediction — yes, and `Daily_Commute_km` is the big one.** Residual (observed − recipe probability), baseline −0.0162 everywhere:

| commute band (km) | n | residual | buy rate |
|---|---|---|---|
| = 5.0 (floor) | 144,280 | −0.0145 | 0.1844 |
| **(10, 15]** | **9,024** | **+0.0384** | **0.3020** |
| (15, 20] | 29,031 | +0.0062 | 0.2348 |
| (20, 40] | 226,556 | −0.0180 | 0.1749 |
| (40, 60] | 226,058 | −0.0186 | 0.1649 |
| **(60, 70]** | **27,967** | **−0.0334** | **0.1066** |

Not a confound: `corr(commute, income) = 0.008`, `corr(commute, concern) = −0.019`. And it **survives conditioning on the recipe score** — buy rate (%) by score decile × commute band:

| decile | ≤5 | (5,20] | (20,40] | (40,60] | >60 |
|---|---|---|---|---|---|
| 6 | 9.72 | **12.71** | 9.26 | 9.50 | 7.68 |
| 7 | 28.04 | **35.80** | 27.36 | 26.30 | 21.73 |
| 8 | 54.50 | **61.00** | 53.13 | 52.94 | 45.25 |
| 9 | 79.56 | **84.19** | 77.96 | 77.66 | 72.77 |

A consistent ~+6 pp lift for 5–20 km and ~−6 pp for >60 km inside *every* score decile. `Age` shows a smaller clean bump (50–55 band: residual −0.0032, buy rate 0.2030 vs a 0.1746 base). City type, home charging, charger counts, car type, gender and cars owned are all within ±0.01 of baseline, |Spearman| < 0.03. The recipe also over-predicts monotonically across the whole upper score range (ventile 16: 0.384 predicted vs 0.327 observed).

### Decisions

**Recipe as `init_score` / `base_margin` — NO for turn 1.** Measured in our own frame (fold 0, `metric='auc'`, early stopping 200; no-init baseline 0.940414):

| variant | best_iter | AUC | delta |
|---|---|---|---|
| none | 434 | 0.940414 | — |
| logit of recipe prob, clip 1e-6 | 132 | 0.937663 | **−0.00275** |
| logit, clip 1e-4 | 243 | 0.937865 | −0.00255 |
| logit, clip 1e-3 | 345 | 0.937940 | −0.00247 |
| logit, clip 1e-3, shrunk ×0.5 | 344 | 0.938566 | −0.00185 |
| raw margin `(s−5.5)/0.802` | 337 | 0.938420 | −0.00199 |
| full 5-fold OOF, clip 1e-6 | — | 0.938763 | −0.00290 vs 0.941665 |

Every variant **hurts**, by 0.002–0.003, robust to clipping and shrinkage; `best_iter` collapses because an anchored margin saturates the extreme score regions and the booster can no longer recover the within-region income fine structure that supplies most of its edge. Even at face value the field's **+0.00005–0.00007 is below the 0.0002 LB noise floor**. As a plain column we measured **+0.00004** (0.941665 → 0.941704), inside noise — matching Deotte. Not worth a moving part in turn 1's baseline; revisit only if turn 2 reaches a plateau-class frame with spare slots.

**Original 10,000 rows as extra training data — NO, rule it out.** 1.5% row bump, matching distribution, field-measured label-joining at **+0.00003** (an order of magnitude under the noise floor) and income × recipe-column keying measured at **−0.0004**. Our own numbers reinforce it: the value the original rows carry is the exact-income identity, and that is already extractable from our 668k rows by target/count encoding of income with no join at all.

**Reproducing the generator from seed 101 — NO, take the formula as given.** Two numbers settle it: the free refit on our own 668k rows buys **+0.00008** over the published constants (so the recipe is already correct and a byte-exact reproduction cannot sharpen it), and the only extra artefact a reproduction yields — original row identity — is measured at **+0.00003**. A multi-hour task with a ceiling of 0.00003–0.00008, both under the noise floor, against an 8-day budget. Empirical verification against our train set, which this ticket did, is sufficient.

### What this changes downstream

1. **The ceiling is not arithmetic.** The recipe accounts for 0.9377 of the ~0.9466 plateau; the remaining 0.0089 is recoverable structure, most of it (+0.0029) non-linear use of the recipe's *own* inputs. This confirms digit decomposition / income target encoding / high `max_bin` as the priority feature work and explains the mechanism.
2. **`Daily_Commute_km` is the most promising non-recipe feature and is currently under-rated** — #2's single-feature AUC of 0.534 hides a non-monotone effect worth ~6 pp of buy rate inside every recipe-score decile. Bin it, or ensure the model can split it finely; the 5–20 km and >60 km bands are the two edges.
3. **Do not add recipe-derived quantities to turn 1 in any form.** A GBDT rediscovers the recipe from 669k rows on its own.

**Comment by ronaldoi9** [comment]
## Resolution

**The hypothesis this ticket was built on is refuted.** Full numbers in the comment above and in `docs/research/generator-recipe.md` on branch `research/generator-recipe`.

**The central measurement.** Scoring all 668,665 train rows with the published noiseless formula and taking AUC directly against `Will_Buy_EV`:

| | AUC |
|---|---|
| Published formula, noiseless, as a ranking | **0.937690** |
| Free-coefficient logistic refit on the same 4 variables | 0.937772 |
| Published plateau | ~0.9464 CV / 0.9466 LB |

**Gap: -0.0089, about 45x the leaderboard noise floor.** The reasoning was sound — additive independent noise makes `P(y=1|x)` monotone in `buy_score`, so the noiseless score is the Bayes ranking under the published model — but that model's Bayes ranking is 0.9377, far under what the field reaches. **The residual positives are not noise. The ceiling is a modelling problem, not arithmetic.**

**The recipe itself is confirmed.** A free refit renormalised to `concern = 0.6` recovers income 1.157 (published 1.2), subsidy 2.072 (2), Medium -0.841 (-1), High -1.728 (-3, weakly identified at 0.33% of rows), threshold 5.647 (5.5).

**Where the missing 0.0089 lives** (5-fold LightGBM, plain frame, no digit or target-encoding features):

| model | OOF AUC | delta |
|---|---|---|
| Formula as a score, no model | 0.937690 | — |
| LightGBM on the 4 recipe variables only | 0.940632 | **+0.00294** |
| LightGBM on all 13 features | 0.941665 | +0.00103 |

**Two thirds of the gap is non-linear structure inside the recipe's own four variables** — the exact-income artifact seen from the other side. Only +0.0010 comes from the nine features the formula omits. The arithmetic closes: 0.9377 + 0.0029 + 0.0010 = 0.9417, plus the published ~0.0038 of feature engineering, plus tuning, reaches the plateau. Interpretation: the recipe is the label rule of the **original 10k**; our 668k were regenerated by a learned model that only approximately preserves it.

**Sanity checks**
1. **Positive rate**: the noiseless `s > 5.5` rule flags 17.794% against an observed 17.4645%. Symmetric zero-mean noise only raises that (infimum 0.1778), so **no noise scale reproduces the observed rate at threshold 5.5** — it needs ~5.65, exactly what the refit found. Inferred noise: Gaussian sigma 0.802 (MLE); plain N(0,1) rejected.
2. **Coefficients against single-feature AUCs**: consistent. Term SDs subsidy 0.967 / concern 0.858 / income 0.344 / anxiety 0.335 against AUCs 0.718 / 0.844 / 0.670 / 0.545, once binary-AUC capping and High's rarity are accounted for.
3. **Systematic misprediction found: `Daily_Commute_km`.** Residual **+0.0384** for the 10-15 km band (buy rate 0.3020 against a 0.1746 base) and **-0.0334** for 60-70 km (0.1066). Not a confound (correlation 0.008 with income, -0.019 with concern) and it **survives conditioning on the recipe score — roughly +6pp for 5-20 km and -6pp above 60 km inside every score decile**. `Age` 50-55 shows a smaller clean bump (0.2030). Everything else is within +/-0.01.

**Three decisions, each with its number**

- **`init_score` / `base_margin`: no, not for turn 1.** Measured **negative** in our own frame — every variant hurts: clip 1e-6 -0.00275, clip 1e-4 -0.00255, shrunk x0.5 -0.00185, raw margin -0.00199, full 5-fold -0.00290 against a fold-0 baseline of 0.940414. `best_iter` collapses from 434 to 130-345: the anchored margin saturates the extreme score regions and blocks the income fine structure. As a plain column it is +0.00004, inside noise. Even the field's own reported +0.00005-0.00007 sits below the 0.0002 leaderboard floor.
- **The original 10,000 rows as training data: ruled out.** Field-measured at +0.00003 for label joining and -0.0004 for income x recipe keying, and the exact-income identity they carry is already extractable from our own 668k by target or count encoding with no join at all.
- **Reproducing the generator from seed 101: no, take the formula as given.** A free refit buys +0.00008 over the published constants, and the only extra artifact is worth +0.00003. Multi-hour task with a ceiling under the noise floor.

**Carried forward**
1. The priority feature work now has a mechanism, not just an ablation table: digit decomposition, income target encoding and a high `max_bin` matter **because +0.0029 of the gap is non-linear structure in the recipe's own inputs**.
2. **`Daily_Commute_km` is under-rated.** Its single-feature AUC of 0.534 hides a non-monotone effect worth ~6pp inside every score decile. Bin it or ensure fine splits; 5-20 km and >60 km are the edges.
